apps/web/vibes/soul/sections/cart/coupon-code-form.tsx

```diff
@@ -0,0 +1,109 @@
+'use client';
+
+import { getFormProps, getInputProps, SubmissionResult, useForm } from '@conform-to/react';
+import { parseWithZod } from '@conform-to/zod';
+import { useActionState } from 'react';
+import { useFormStatus } from 'react-dom';
+
+import { FieldError } from '@/vibes/soul/form/field-error';
+import { Input } from '@/vibes/soul/form/input';
+import { Button } from '@/vibes/soul/primitives/button';
+import { Chip } from '@/vibes/soul/primitives/chip';
+
+import { couponCodeActionFormDataSchema } from './schema';
+
+type Action<State, Payload> = (state: Awaited<State>, payload: Payload) => State | Promise<State>;
+
+export interface CouponCodeFormState {
+  couponCodes: string[];
+  lastResult: SubmissionResult | null;
+}
+
+interface Props {
+  action: Action<CouponCodeFormState, FormData>;
+  couponCodes?: string[];
+  ctaLabel?: string;
+  disabled?: boolean;
+  label?: string;
+  placeholder?: string;
+  removeLabel?: string;
+}
+
+export function CouponCodeForm({
+  action,
+  couponCodes,
+  ctaLabel = 'Apply',
+  disabled = false,
+  label = 'Promo code',
+  placeholder,
+  removeLabel = 'Remove promo code',
+}: Props) {
+  const [{ couponCodes: actionStateCouponCodes, lastResult }, formAction] = useActionState(action, {
+    couponCodes: couponCodes ?? [],
+    lastResult: null,
+  });
+
+  const [form, fields] = useForm({
+    lastResult,
+    shouldValidate: 'onBlur',
+    shouldRevalidate: 'onInput',
+    onValidate({ formData }) {
+      return parseWithZod(formData, { schema: couponCodeActionFormDataSchema });
+    },
+  });
+
+  return (
+    <div className="border-t border-contrast-100 py-6">
+      <form {...getFormProps(form)} action={formAction} className="space-y-2">
+        <label htmlFor={fields.couponCode.id}>{label}</label>
+        <div className="flex gap-1.5">
+          <Input
+            {...getInputProps(fields.couponCode, {
+              required: true,
+              type: 'text',
+            })}
+            disabled={disabled}
+            errors={fields.couponCode.errors}
+            id={fields.couponCode.id}
+            key={fields.couponCode.id}
+            placeholder={placeholder}
+          />
+          <SubmitButton disabled={disabled}>{ctaLabel}</SubmitButton>
+        </div>
+        {form.errors?.map((error, index) => <FieldError key={index}>{error}</FieldError>)}
+      </form>
+      <div className="flex flex-wrap gap-1.5 pt-3">
+        {actionStateCouponCodes.map((couponCode) => (
+          <form action={formAction} key={couponCode}>
```

Comment:
Using the same action with `intent`. For these forms, I'm not using `getFormProps`, etc, seems to be only useful for forms with inputs, but I'm not convinced if this is the right way.

Reply:
How come we have two forms, though? We should only have 1. If we have 1 then we don't need to use `getFormProps` here